README.md
# SilverStripe Geocoder

SilverStripe wrapper for [Geocoder](https://github.com/geocoder-php/Geocoder)

[![CI](https://github.com/dynamic/silverstripe-geocoder/actions/workflows/ci.yml/badge.svg)](https://github.com/dynamic/silverstripe-geocoder/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/dynamic/silverstripe-geocoder/branch/master/graph/badge.svg)](https://codecov.io/gh/dynamic/silverstripe-geocoder)

[![Latest Stable Version](https://poser.pugx.org/dynamic/silverstripe-geocoder/v/stable)](https://packagist.org/packages/dynamic/silverstripe-geocoder)
[![Total Downloads](https://poser.pugx.org/dynamic/silverstripe-geocoder/downloads)](https://packagist.org/packages/dynamic/silverstripe-geocoder)
[![Latest Unstable Version](https://poser.pugx.org/dynamic/silverstripe-geocoder/v/unstable)](https://packagist.org/packages/dynamic/silverstripe-geocoder)
[![License](https://poser.pugx.org/dynamic/silverstripe-geocoder/license)](https://packagist.org/packages/dynamic/silverstripe-geocoder)

## Requirements

- Silverstripe ^5
- dynamic/silverstripe-country-dropdown-field ^2
- geocoder-php/google-maps-provider ^4.7
- guzzlehttp/guzzle ^7.4
- php-http/guzzle7-adapter ^1.0
- php-http/message ^1.13

## Installation

`composer require dynamic/silverstripe-geocoder`

## License

See [License](license.md)

## Example usage

### Getting Started

in `mysite/_config/config.yml`, apply the DataExtensions to your DataObject:
```yaml
SilverStripe\ORM\DataObject:
  extensions:
    - Dynamic\SilverStripeGeocoder\AddressDataExtension
    - Dynamic\SilverStripeGeocoder\DistanceDataExtension
```

### Google API Keys

You'll also need to set two [Google API keys](https://developers.google.com/maps/documentation/javascript/get-api-key). Each key needs to have specific API libraries enabled:
* `geocoder_api_key`
  * Geocoding API
* `map_api_key`
  * Maps JavaScript API
  *  Maps Static API

This is primarily due to the restriction rules on Google API keys. The Geocoding key should be tied to the webserver's public IP, while the Maps API needs to be restricted to the website's domain name.
Restriction rules help to prevent unauthorized users to use your keys, which can quickly use up your API requests quota.

```yaml
Dynamic\SilverStripeGeocoder\GoogleGeocoder:
  geocoder_api_key: 'your-key-here'
  map_api_key: 'your-key-here'
```

## Documentation

### AddressDataExtension
The AddressDataExtension adds `Address`, `Address2`, `City`, `State`, `PostalCode`, `Country`, `Lat`, and `Lng` fields.
The `Lat` and `Lng` fields are read only in the cms and are automatically generated on write.
Geocoding can be disabled on a model basis by setting `disable_geocoding` to true.

```yml
SilverStripe\ORM\DataObject:
    disable_geocoding: true
```

#### Static Map Image

Using `$AddressMap` in a template will render the map.
`$AddressMap` also has options to easily modify the width, height, and scale of the map. `$AddressMap(320, 240, 1)`

##### Map Style

The map can be styled by adding a `mapStyle.json` in any of the following folders in a theme:
```
client/dist/js/
client/dist/javascript/
dist/js/
dist/javascript/
src/javascript/thirdparty
js/
javascript/
```

Styles can be generated using an online service like [Styling Wizard: Google Maps APIs](https://mapstyle.withgoogle.com/)

##### Marker Image
**This does not work on local due to google needing to download the image off the server.**

A custom marker image can be used to match the style of the map in about the same way as the style.
An image named `mapIcon` with an extension of `png`, `jpg`, `jpeg`, or `gif` can be put in any of the following folders in a theme:
```
client/dist/img/
client/dist/images/
dist/img/
dist/images/
img/
images/
```

### DistanceDataExtension
The DistanceDataExtension should be used in conjunction with the AddressDataExtension.
The only time it is viable by itself is if the extended DataObject has `Lat` and `Lng` fields.

The DistanceDataExtension will add a pseudo field for distance away from an address to a DataObject.
The address to check the distance to is from the current controller initially.
This can be changed by implementing `updateAddressValue($address)` on the DataObject or an extension.

```php
/**
 * Always get the distance from Neuschwanstein Castle
 */
public function updateAddressValue(&$address) {
    $address = 'Neuschwansteinstraße 20, 87645 Schwangau, Germany';
}
```

## Maintainers
 *  [Dynamic](http://www.dynamicagency.com) (<dev@dynamicagency.com>)

## Bugtracker
Bugs are tracked in the issues section of this repository. Before submitting an issue please read over
existing issues to ensure yours is unique.

If the issue does look like a new bug:

 - Create a new issue
 - Describe the steps required to reproduce your issue, and the expected outcome. Unit tests, screenshots
 and screencasts can help here.
 - Describe your environment as detailed as possible: SilverStripe version, Browser, PHP version,
 Operating System, any installed SilverStripe modules.

Please report security issues to the module maintainers directly. Please don't file security issues in the bugtracker.

## Development and contribution
If you would like to make contributions to the module please ensure you raise a pull request and discuss with the module maintainers.


CHANGELOG.md
# Changelog

## [3.0.0](https://github.com/dynamic/silverstripe-geocoder/tree/3.0.0) (2023-08-26)
* REFACTOR Silverstripe 5 compatibility by @jsirish in https://github.com/dynamic/silverstripe-geocoder/pull/70

**Full Changelog**: https://github.com/dynamic/silverstripe-geocoder/compare/2.0.4...3.0.0

## [2.0.4](https://github.com/dynamic/silverstripe-geocoder/tree/2.0.4) (2023-08-26)
* CI workflow to SS GHA by @jsirish in https://github.com/dynamic/silverstripe-geocoder/pull/68
* REFACTOR README additions by @jsirish in https://github.com/dynamic/silverstripe-geocoder/pull/67
* REFACTOR allow country dropdown field 1 or 2 by @jsirish in https://github.com/dynamic/silverstripe-geocoder/pull/69


**Full Changelog**: https://github.com/dynamic/silverstripe-geocoder/compare/2.0.3...2.0.4

## [1.3.0](https://github.com/dynamic/silverstripe-geocoder/tree/1.3.0) (2022-07-26)

## What's Changed
* REFACTOR PHP8 compatibility by @jsirish in https://github.com/dynamic/silverstripe-geocoder/pull/55


**Full Changelog**: https://github.com/dynamic/silverstripe-geocoder/compare/1.2.0...1.3.0

## [1.2.0](https://github.com/dynamic/silverstripe-geocoder/tree/1.2.0) (2020-08-03)

[Full Changelog](https://github.com/dynamic/silverstripe-geocoder/compare/1.1.1...1.2.0)

**Closed issues:**

- FEATURE Allow user to override lat and lng [\#53](https://github.com/dynamic/silverstripe-geocoder/issues/53)

**Merged pull requests:**

- FEATURE LatLngOverride [\#54](https://github.com/dynamic/silverstripe-geocoder/pull/54) ([jsirish](https://github.com/jsirish))

## [1.1.1](https://github.com/dynamic/silverstripe-geocoder/tree/1.1.1) (2020-01-09)

[Full Changelog](https://github.com/dynamic/silverstripe-geocoder/compare/1.1.0...1.1.1)

**Implemented enhancements:**

- REFACTOR Geocoding - don't attempt to generate coordinates if geocoding key is missing [\#47](https://github.com/dynamic/silverstripe-geocoder/issues/47)

**Merged pull requests:**

- BUGFIX AddressDataExtension onBeforeWrite - check if geocoding key is… [\#52](https://github.com/dynamic/silverstripe-geocoder/pull/52) ([jsirish](https://github.com/jsirish))

## [1.1.0](https://github.com/dynamic/silverstripe-geocoder/tree/1.1.0) (2019-11-19)

[Full Changelog](https://github.com/dynamic/silverstripe-geocoder/compare/1.0.2...1.1.0)

**Implemented enhancements:**

- REFACTOR Remove StateDropdownField as a default in CMS [\#45](https://github.com/dynamic/silverstripe-geocoder/issues/45)

**Fixed bugs:**

- REFACTOR AddressDataExtension::hasAddress\(\) - allow for State, PostalCode or Country [\#46](https://github.com/dynamic/silverstripe-geocoder/issues/46)

**Closed issues:**

- 1.2.0 [\#48](https://github.com/dynamic/silverstripe-geocoder/issues/48)
- FEATURE Github Issue templates [\#44](https://github.com/dynamic/silverstripe-geocoder/issues/44)

**Merged pull requests:**

- REFACTOR AddressDataExtension - remove StateDropdownField [\#51](https://github.com/dynamic/silverstripe-geocoder/pull/51) ([jsirish](https://github.com/jsirish))
- DOCS add Github templates [\#50](https://github.com/dynamic/silverstripe-geocoder/pull/50) ([jsirish](https://github.com/jsirish))
- REFACTOR AddressDataExtension::hasAddress\(\) [\#49](https://github.com/dynamic/silverstripe-geocoder/pull/49) ([jsirish](https://github.com/jsirish))

## [1.0.2](https://github.com/dynamic/silverstripe-geocoder/tree/1.0.2) (2019-10-29)

[Full Changelog](https://github.com/dynamic/silverstripe-geocoder/compare/1.0.1...1.0.2)

**Fixed bugs:**

- BUG Coordinates are not calculated if Location is missing postal code. Not all cities have postal codes [\#42](https://github.com/dynamic/silverstripe-geocoder/issues/42)

**Merged pull requests:**

- BUGFIX AddressDataExtension - hasAddress\(\) [\#43](https://github.com/dynamic/silverstripe-geocoder/pull/43) ([jsirish](https://github.com/jsirish))

## [1.0.1](https://github.com/dynamic/silverstripe-geocoder/tree/1.0.1) (2019-08-27)

[Full Changelog](https://github.com/dynamic/silverstripe-geocoder/compare/1.0.0...1.0.1)

**Implemented enhancements:**

- Tag a release [\#35](https://github.com/dynamic/silverstripe-geocoder/issues/35)

**Closed issues:**

- FEATURE Location - implement Address 2 field [\#37](https://github.com/dynamic/silverstripe-geocoder/issues/37)
- Note somewhere in the docs that markers for static maps don't work on private sites [\#32](https://github.com/dynamic/silverstripe-geocoder/issues/32)
- Add documentation for model level disabling of geocoding [\#31](https://github.com/dynamic/silverstripe-geocoder/issues/31)

**Merged pull requests:**

- FEATURE AddressDataExtension - CountryDropdownField [\#41](https://github.com/dynamic/silverstripe-geocoder/pull/41) ([jsirish](https://github.com/jsirish))
- Update CI tests [\#39](https://github.com/dynamic/silverstripe-geocoder/pull/39) ([jsirish](https://github.com/jsirish))
- FEATURE AddressDataExtension - Address2 label [\#38](https://github.com/dynamic/silverstripe-geocoder/pull/38) ([jsirish](https://github.com/jsirish))
- Changed front end maps to use a different api key [\#36](https://github.com/dynamic/silverstripe-geocoder/pull/36) ([mak001](https://github.com/mak001))
- ENHANCEMENT state and country dropdown fields [\#20](https://github.com/dynamic/silverstripe-geocoder/pull/20) ([muskie9](https://github.com/muskie9))

## [1.0.0](https://github.com/dynamic/silverstripe-geocoder/tree/1.0.0) (2019-05-08)

[Full Changelog](https://github.com/dynamic/silverstripe-geocoder/compare/05d0dd93d2b1d28ad4fb46a5bba28c458eaa5423...1.0.0)

**Implemented enhancements:**

- Migrate AddressDataExtension and DistanceDataExtension from Locator [\#4](https://github.com/dynamic/silverstripe-geocoder/issues/4)
- FEATURE Use dynamic/silverstripe-country-dropdown-field for Country field [\#40](https://github.com/dynamic/silverstripe-geocoder/issues/40)

tests/TestLocation.php
<?php

namespace Dynamic\SilverStripeGeocoder\TestOnly;

use Dynamic\SilverStripeGeocoder\AddressDataExtension;
use SilverStripe\Dev\TestOnly;
use SilverStripe\ORM\DataObject;

class TestLocation extends DataObject implements TestOnly
{
    /**
     * @var array
     */
    private static $extensions = [
        AddressDataExtension::class,
    ];
}


tests/AddressDataExtensionTest.php
<?php

namespace Dynamic\SilverStripeGeocoder\Tests;

use Dynamic\SilverStripeGeocoder\AddressDataExtension;
use Dynamic\SilverStripeGeocoder\TestOnly\TestLocation;
use SilverStripe\Dev\SapphireTest;
use SilverStripe\Forms\FieldList;

class AddressDataExtensionTest extends SapphireTest
{
    /**
     * @var string
     */
    protected static $fixture_file = 'fixtures.yml';

    /**
     * @var array
     */
    protected static $extra_dataobjects = [
        TestLocation::class,
    ];

    /**
     * @var \string[][]
     */
    protected static $required_extensions = [
        TestLocation::class => [
            AddressDataExtension::class,
        ],
    ];

    /**
     *
     */
    public function testUpdateCMSFields()
    {
        $object = $this->objFromFixture(TestLocation::class, 'dynamic');
        $fields = $object->getCMSFields();
        $this->assertInstanceOf(FieldList::class, $fields);
    }

    /**
     *
     */
    public function testGetMapStyleJSON()
    {
        $object = $this->objFromFixture(TestLocation::class, 'dynamic');
        $this->assertFalse($object->getMapStyleJSON());
    }

    /**
     *
     */
    public function testGetIconImage()
    {
        $object = $this->objFromFixture(TestLocation::class, 'dynamic');
        $this->assertFalse($object->getIconImage());
    }

    /**
     *
     */
    public function testMapStylesUrlArgs()
    {
        $object = $this->objFromFixture(TestLocation::class, 'dynamic');
        $json = '[{"elementType":"geometry","stylers":[{"color":"#ebe3cd"}]},{"featureType":"administrative",
            "elementType":"geometry","stylers":[{"visibility":"off"}]},{"featureType":"administrative","elementType":
            "geometry.stroke","stylers":[{"color":"#c9b2a6"}]}]';
        $args = $object->mapStylesURLArgs($json);

        $expected = 'style=feature:all|element:geometry|color:0xebe3cd&style=feature:administrative|element:geometry|visibility:off&style=feature:administrative|element:geometry.stroke|color:0xc9b2a6';
        $this->assertEquals($expected, $args);
    }

    /**
     *
     */
    public function testGetFullAddress()
    {
    }

    /**
     *
     */
    public function testHasAddress()
    {
    }
}


src/GoogleGeocoder.php
<?php

namespace Dynamic\SilverStripeGeocoder;

use Geocoder\Provider\GoogleMaps\GoogleMaps;
use \SilverStripe\Core\Config\Config;

/**
 * Class GoogleGeocoder
 * @package Dynamic\SilverStripeGeocoder
 */
class GoogleGeocoder
{
    /**
     * @var
     */
    private $client;

    /**
     * @var
     */
    private $results;

    /**
     * GoogleGeocoder constructor.
     * @param $address
     */
    public function __construct($address)
    {
        $httpClient  = new GeocoderAdapter();
        $httpClient = $httpClient->getAdapter();
        $provider = new \Geocoder\Provider\GoogleMaps\GoogleMaps(
            $httpClient,
            null,
            Config::inst()->get(GoogleGeocoder::class, 'geocoder_api_key')
        );
        $geocoder = new \Geocoder\StatefulGeocoder($provider, 'en');

        $this->setClient($geocoder);
        $this->setResults($address);
    }

    /**
     * @return mixed
     */
    public function getClient()
    {
        return $this->client;
    }

    /**
     * @param $client
     * @return $this
     */
    public function setClient($client)
    {
        $this->client = $client;
        return $this;
    }

    /**
     * @return mixed
     */
    public function getResults()
    {
        return $this->results;
    }

    /**
     * @param $address
     * @return $this
     */
    public function setResults($address)
    {
        $client = $this->getClient();
        $this->results = $client->geocode($address);
        return $this;
    }

    /**
     * @return mixed
     */
    public function getResult()
    {
        return $this->getResults()->get(0);
    }
}


src/DistanceDataExtension.php
<?php

namespace Dynamic\SilverStripeGeocoder;

use SilverStripe\Core\Config\Config;
use SilverStripe\ORM\DataExtension;
use SilverStripe\ORM\Queries\SQLSelect;
use SilverStripe\ORM\DataQuery;
use SilverStripe\Control\Controller;

/**
 * Class \Dynamic\SilverStripeGeocoder\DistanceDataExtension
 *
 * @property DistanceDataExtension $owner
 */
class DistanceDataExtension extends DataExtension
{
    /**
     * @param SQLSelect $query
     * @param DataQuery|null $dataQuery
     */
    public function augmentSQL(SQLSelect $query, DataQuery $dataQuery = null)
    {
        $addressVar = Config::inst()->get(DistanceDataExtension::class, 'address_var');
        $unitVar = Config::inst()->get(DistanceDataExtension::class, 'unit_var');

        $address = Controller::curr()->getRequest()->getVar($addressVar);
        if ($this->owner->hasMethod('updateAddressValue')) {
            $address = $this->owner->updateAddressValue($address);
        }

        // switch between miles and kilometers
        $unit = Controller::curr()->getRequest()->getVar($unitVar);
        if ($unit == 'km') {
            $unitVal = 6371;
        } else {
            $unitVal = 3959;
        }

        if ($address) { // on frontend
            $geocoder = new GoogleGeocoder($address);
            $response = $geocoder->getResult();
            $Lat = $this->owner->Lat = $response->getCoordinates()->getLatitude();
            $Lng = $this->owner->Lng = $response->getCoordinates()->getLongitude();

            $query
                ->addSelect(array(
                    '( ' . $unitVal . ' * acos( cos( radians(' . $Lat .
                    ') ) * cos( radians( `Lat` ) ) * cos( radians( `Lng` ) - radians(' . $Lng . ') ) + sin( radians(' .
                    $Lat . ') ) * sin( radians( `Lat` ) ) ) ) AS Distance',
                ));
        } else {
            $query->addSelect(array('-1 AS Distance'));
        }
    }
}


src/AddressDataExtension.php
<?php

namespace Dynamic\SilverStripeGeocoder;

use Dynamic\CountryDropdownField\Fields\CountryDropdownField;
use SilverStripe\Control\Director;
use Dynamic\StateDropdownField\Fields\StateDropdownField;
use SilverStripe\Core\Config\Config;
use SilverStripe\Core\Manifest\ModuleResourceLoader;
use SilverStripe\Forms\CheckboxField;
use SilverStripe\Forms\CompositeField;
use SilverStripe\Forms\FieldList;
use SilverStripe\Forms\LiteralField;
use SilverStripe\Forms\ReadonlyField;
use SilverStripe\Forms\TextField;
use SilverStripe\ORM\DataExtension;
use SilverStripe\View\SSViewer;
use SilverStripe\View\ThemeResourceLoader;

/**
 * Class \Dynamic\SilverStripeGeocoder\AddressDataExtension
 *
 * @property CompanyAddress|AddressDataExtension $owner
 * @property string $Address
 * @property string $Address2
 * @property string $City
 * @property string $State
 * @property string $PostalCode
 * @property string $Country
 * @property bool $LatLngOverride
 * @property float $Lat
 * @property float $Lng
 */
class AddressDataExtension extends DataExtension
{
    /**
     * @var array
     */
    private static $db = [
        'Address' => 'Varchar(255)',
        'Address2' => 'Varchar(255)',
        'City' => 'Varchar(64)',
        'State' => 'Varchar(64)',
        'PostalCode' => 'Varchar(10)',
        'Country' => 'Varchar(2)',
        'LatLngOverride' => 'Boolean',
        'Lat' => 'Decimal(10,7)',
        'Lng' => 'Decimal(10,7)',
    ];

    /**
     * @param FieldList $fields
     */
    public function updateCMSFields(FieldList $fields)
    {
        if ($this->owner->config()->address_tab_name) {
            $tab_name = $this->owner->config()->address_tab_name;
        } else {
            $tab_name = Config::inst()->get(AddressDataExtension::class, 'address_tab_name');
        }

        $fields->removeByName([
            'LatLngOverride',
            'Lat',
            'Lng',
        ]);

        $fields->addFieldsToTab('Root.' . $tab_name, [
            TextField::create('Address'),
            TextField::create('Address2', 'Address 2'),
            TextField::create('City'),
            TextField::create('State', 'State/Province'),
            TextField::create('PostalCode'),
            CountryDropdownField::create('Country')
                ->setEmptyString('Select Country'),
        ]);

        $compositeField = CompositeField::create();
        $compositeField->push($overrideField = CheckboxField::create('LatLngOverride', 'Override Latitude and Longitude?'));
        $overrideField->setDescription('Check this box and save to be able to edit the latitude and longitude manually.');

        if ($this->owner->Lng && $this->owner->Lat) {
            $googleMapURL = 'https://maps.google.com/?q=' . $this->owner->Lat . ',' . $this->owner->Lng;
            $googleMapDiv = '<div class="field"><label class="left" for="Form_EditForm_MapURL_Readonly">Google Map</label><div class="middleColumn"><a href="' . $googleMapURL . '" target="_blank">' . $googleMapURL . '</a></div></div>';
            $compositeField->push(LiteralField::create('MapURL_Readonly', $googleMapDiv));
        }
        if ($this->owner->LatLngOverride) {
            $compositeField->push(TextField::create('Lat', 'Lat'));
            $compositeField->push(TextField::create('Lng', 'Lng'));
        } else {
            $compositeField->push(ReadonlyField::create('Lat_Readonly', 'Lat', $this->owner->Lat));
            $compositeField->push(ReadonlyField::create('Lng_Readonly', 'Lng', $this->owner->Lng));
        }

        $fields->addFieldToTab('Root.' . $tab_name, $compositeField);
    }

    /**
     * Returns the full address as a simple string.
     *
     * @return string
     */
    public function getFullAddress()
    {
        $parts = [
            $this->owner->Address,
            $this->owner->Address2,
            $this->owner->City,
            $this->owner->State,
            $this->owner->PostalCode,
        ];

        if ($this->owner->Country !== null) {
            $parts[] = strtoupper($this->owner->Country);
        }

        return implode(', ', array_filter($parts));
    }

    /**
     * @return bool
     */
    public function hasAddress()
    {
        $address =
            $this->owner->Address
            && $this->owner->City
            && ($this->owner->State || $this->owner->PostalCode || $this->owner->Country);

        $this->owner->extend('updateHasAddress', $address);

        return $address;
    }

    /**
     * Returns a static google map of the address, linking out to the address.
     *
     * @param int $width (optional)
     * @param int $height (optional)
     * @param int $scale (optional)
     *
     * @return string
     */
    public function AddressMap($width = 320, $height = 240, $scale = 1)
    {
        $styleJSON = static::getMapStyleJSONPath();
        $style = false;
        if ($styleJSON !== false) {
            $style = $this->mapStylesUrlArgs(file_get_contents($styleJSON));
        }

        $icon = Director::absoluteURL(static::getIconImage(false));

        $data = $this->owner->customise([
            'Width' => $width,
            'Height' => $height,
            'Scale' => $scale,
            'Address' => rawurlencode($this->getFullAddress()),
            'Style' => $style,
            'Icon' => $icon,
            'Key' => Config::inst()->get(GoogleGeocoder::class, 'map_api_key'),
        ]);

        return $data->renderWith('Dynamic/Geocoder/AddressMap');
    }

    /**
     * @return bool|string
     */
    public static function getMapStyleJSONPath()
    {
        if ($styleJSON = static::getMapStyleJSON()) {
            return BASE_PATH . DIRECTORY_SEPARATOR . $styleJSON;
        }

        return false;
    }

    /**
     * Gets the style of the map
     * @return string|null
     */
    public static function getMapStyleJSON()
    {
        $folders = [
            'client/dist/js/',
            'client/dist/javascript/',
            'dist/js/',
            'dist/javascript/',
            'src/javascript/thirdparty',
            'js/',
            'javascript/',
        ];
        $file = 'mapStyle.json';

        foreach ($folders as $folder) {
            if ($style = ThemeResourceLoader::inst()->findThemedResource(
                "{$folder}{$file}",
                SSViewer::get_themes()
            )) {
                return $style;
            }
        }

        return false;
    }

    /**
     * Gets the maker icon image
     * @return null|string
     * @var boolean $svg if svgs should be included
     */
    public static function getIconImage($svg = true)
    {
        $folders = [
            'client/dist/img/',
            'client/dist/images/',
            'dist/img/',
            'dist/images/',
            'img/',
            'images/',
        ];

        $extensions = [
            'png',
            'jpg',
            'jpeg',
            'gif',
        ];

        if ($svg === true) {
            array_unshift($extensions, 'svg');
        }

        $file = 'mapIcon';

        foreach ($folders as $folder) {
            foreach ($extensions as $extension) {
                if ($icon = ThemeResourceLoader::inst()->findThemedResource(
                    "{$folder}{$file}.{$extension}",
                    SSViewer::get_themes()
                )) {
                    return ModuleResourceLoader::resourceURL($icon);
                }
            }
        }

        return false;
    }

    /**
     * From https://gist.github.com/wouterds/5942b891cdad4fc90f40
     * @param $mapStyleJson
     *
     * @return string
     */
    public function mapStylesUrlArgs($mapStyleJson)
    {
        $params = [];
        foreach (json_decode($mapStyleJson, true) as $style) {
            $styleString = '';
            if (isset($style['stylers']) && count($style['stylers']) > 0) {
                $styleString .= (isset($style['featureType']) ?
                        ('feature:' . $style['featureType']) : 'feature:all') . '|';
                $styleString .= (isset($style['elementType']) ?
                        ('element:' . $style['elementType']) : 'element:all') . '|';
                foreach ($style['stylers'] as $styler) {
                    $propertyname = array_keys($styler)[0];
                    $propertyval = str_replace('#', '0x', $styler[$propertyname]);
                    $styleString .= $propertyname . ':' . $propertyval . '|';
                }
            }
            $styleString = substr($styleString, 0, strlen($styleString) - 1);
            $params[] = 'style=' . $styleString;
        }

        return implode("&", $params);
    }

    /**
     * Returns TRUE if any of the address fields have changed.
     *
     * @param int $level
     *
     * @return bool
     */
    public function isAddressChanged($level = 1)
    {
        $fields = ['Address', 'Address2', 'City', 'State', 'PostalCode', 'Country'];
        $changed = $this->owner->getChangedFields(false, $level);
        foreach ($fields as $field) {
            if (array_key_exists($field, $changed)) {
                return true;
            }
        }

        return false;
    }

    /**
     *
     */
    public function onBeforeWrite()
    {
        parent::onBeforeWrite();
        if ($this->hasAddress() && !$this->owner->config()->get('disable_geocoding')
            && Config::inst()->get(GoogleGeocoder::class, 'geocoder_api_key')) {
            if (!$this->isAddressChanged()) {
                return;
            }
            if ($this->owner->LatLngOverride) {
                return;
            }
            if ($address = $this->getFullAddress()) {
                $geocoder = new GoogleGeocoder($address);
                $response = $geocoder->getResult();
                $this->owner->Lat = $response->getCoordinates()->getLatitude();
                $this->owner->Lng = $response->getCoordinates()->getLongitude();
            }
        }
    }
}
